state_tracking.go

```diff
@@ -28,30 +30,53 @@ type StateTrackingNode struct {
 	as     string
 
 	newTracker func() stateTracker
-	groups     map[models.GroupID]*stateTrackingGroup
+
+	groupsMu sync.RWMutex
+	groups   map[models.GroupID]*stateTrackingGroup
 }
 
 func (stn *StateTrackingNode) group(g models.GroupID) (*stateTrackingGroup, error) {
+	stn.groupsMu.RLock()
 	stg := stn.groups[g]
-	if stg == nil {
-		stg = &stateTrackingGroup{}
+	stn.groupsMu.RUnlock()
 
-		var err error
-		stg.Expression, err = stateful.NewExpression(stn.lambda.Expression)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to compile expression: %v", err)
-		}
+	if stg == nil {
+		// Grab the write lock
+		stn.groupsMu.Lock()
+		defer stn.groupsMu.Unlock()
+
+		// Check again now that we have the write lock
+		stg = stn.groups[g]
+		if stg == nil {
```

Comment:
@phemmer On a slightly related note, with the work to change everything to be message passing internally (see #1319), then we can also refactor much of the code to have grouping semantics outside of each node. This way most of the locking logic etc can be moved to a common place. Which I believe is what you expected to see when you first dug into the code. In other words we are headed in that direction and sooner rather than later, so that we don't build up too much momentum in the current state.